### PR TITLE
1609 implement tool to extract all related entities for a study

### PR DESCRIPTION
**Pull request by mbthornton-lbl**
#1609 

Provides a click script: `src/scripts/nmdc_database_tool.py` with an `extract-study` command that extracts data for a given study via the runtime API.

Extraction:
- Study data
- all Biosamples that are `part_of` the Study
- all OmicsProcessings that are `part_of` the Study
- for each OmicsProcessing, all WorkflowExecutionActivities that `was_informed_by` that OmicsProcessing
- for each WorkflowExecutionActivity, all DataObjects for which the WorkflowExecutionActivity `has_output`

Includes extracted data for the "Stegen" study, as well as the extraction log

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -47,6 +47,7 @@ click-log = "^0.4.0"
 linkml = "^1.6.3"
 linkml-runtime = "^1.6.2"
 python = "^3.9"
+click = "^8.1.7"
```
Comment (turbomam) [inline pyproject.toml:50]: I don't think this is required, at least not in both the core dependencies and the dev dependencies. I don't see why we would use different version in those two groups either.

`click` is provided even without a explicit request as it is a dependency of linkml and linkml-runtime

File: src/scripts/db_tool.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# nmdc_schema/src/scripts/db_tool.py
+"""
+db_tool.py: Provide command-line tools for interacting with the NMDC database.
```
Comment (turbomam) [inline src/scripts/db_tool.py:5]: let's think of a more explicit name. something that reflects that it is retrieving related records from the database

File: src/scripts/db_tool.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+@cli.command()
+@click.option(
+    "--study_id",
+    default=STEGEN_STUDY_ID,
+    help=f"Optional updated study ID. Default: {STEGEN_STUDY_ID}",
+)
+@click.option("--yaml", is_flag=True, default=False, help=("Output in YAML "
+                                                         "format."))
+@click.pass_context
+def extract_study(ctx, study_id, yaml):
```
Comment (turbomam) [inline src/scripts/db_tool.py:165]: excellent documentation

File: pyproject.toml
Comment (turbomam) [inline pyproject.toml:1]: let's add a `tool.poetry.scripts` alias for this program

Comment (turbomam) [reply pyproject.toml:1]: although that might not work out of `src/scripts`. I'll give it a try.

**Review by turbomam — COMMENTED** [review]
I just naively typed in 

```shell
poetry run python src/scripts/db_tool.py extract-study
```

and got the following. I'll read the source more carefully now.

> ```
> Extracting study nmdc:sty-11-aygzgv51 from the NMDC database.
> Traceback (most recent call last):
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 265, in <module>
>     cli(obj={})
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/decorators.py", line 33, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 186, in extract_study
>     study_response = api_client.request("GET", f"studies/{study_id}")
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 90, in request
>     self.ensure_token()
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 62, in ensure_token
>     self.get_token()
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 77, in get_token
>     self.base_url + "token", data=token_request_body
> TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
> ```
>

File: src/scripts/nmdc-sty-11-aygzgv51.json
Comment (turbomam) [inline src/scripts/nmdc-sty-11-aygzgv51.json:1]: please send output and log to local/ by default, not the directory in which the script lives

**Comment by turbomam** [comment]
Am I responsible for creating the following environment variables before I launch the script? Let's source them from `local/.env`

- NAPA_BASE_URL
- NAPA_USERNAME
- NAPA_PASSWORD

File: pyproject.toml
Comment (turbomam) [reply pyproject.toml:1]: create a `src/scripts/__init__.py` 

add this to `pyproject.toml`:
`get-study-related-records = "src.scripts.db_tool:cli"`

**Comment by turbomam** [comment]
@mbthornton-lbl after the searches I started yesrteday finished, I got this before anything was aritten to dosk:

> ```
> Traceback (most recent call last):
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 269, in <module>
>     cli(obj={})
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/home/mark/gitrepos/nmdc-schema/.venv/lib/python3.10/site-packages/click/decorators.py", line 33, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/home/mark/gitrepos/nmdc-schema/src/scripts/db_tool.py", line 256, in extract_study
>     yaml_data = yaml.load(yaml_dumper.dumps(db), Loader=yaml.FullLoader)
> AttributeError: 'bool' object has no attribute 'load'
> ```

File: pyproject.toml (hunk 1 above)
Comment (mbthornton-lbl) [reply pyproject.toml]: Removed the redundant requirement and updated the lock file

File: src/scripts/db_tool.py (hunk 2 above)
Comment (mbthornton-lbl) [reply src/scripts/db_tool.py]: nmdc_database_tools

**Review by turbomam — CHANGES_REQUESTED** [review]
the documentation generation phase of `make all` doesn't work in this branch anymore, starting with

```shell
poetry run gen-doc -d docs --template-directory src/doc-templates src/schema/nmdc.yaml
```

**Review by turbomam — APPROVED** [review]

**Comment by cmungall** [comment]
Minor: I don't see why this is being added to nmdc-schema rather than runtime

**Comment by turbomam** [comment]
Yes, we need to find one home for the data access, migration and validation code that several people are working on.

I find nmdc-runtime hard to work in so have stayed in my comfort zone.

**Pull request merged**